src/irmin-pack/unix/lower.ml

```diff
@@ -136,6 +126,80 @@ module Make_volume (Io : Io.S) (Errs : Io_errors.S with module Io = Io) = struct
         match sparse with
         | None -> Errs.raise_error `Closed
         | Some s -> Sparse.read_range_exn s ~off ~min_len ~max_len b)
+
+  let archive_seq ~upper_root ~generation ~is_first ~to_archive ~off t =
+    let open Result_syntax in
+    let root = path t in
+    let* () =
+      match t with
+      | Empty _ -> Ok ()
+      | Nonempty { control; _ } ->
+          if control.end_offset = off then Ok ()
+          else Error `Volume_history_discontinuous
+    in
+    let mapping = Irmin_pack.Layout.V5.Volume.mapping ~root in
+    let data = Irmin_pack.Layout.V5.Volume.data ~root in
+    (* If this is a new volume (and not the first),
+       copy pre-GC prefix/mapping as new volume *)
+    let* () =
+      if (not is_first) && is_empty t then
+        let old_generation = pred generation in
+        let old_mapping =
+          Irmin_pack.Layout.V5.mapping ~root:upper_root
+            ~generation:old_generation
+        in
+        let old_prefix =
+          Irmin_pack.Layout.V5.prefix ~root:upper_root
+            ~generation:old_generation
+        in
+        let* () = Io.copy_file ~src:old_prefix ~dst:data in
+        Io.copy_file ~src:old_mapping ~dst:mapping
+      else Ok ()
+    in
+    let mapping_size =
+      match t with
+      | Empty _ ->
+          (* should this be the size of old_mapping in the above case? *)
+          Int63.zero
+      | Nonempty { control; _ } -> control.mapping_end_poff
+    in
+    (* Append archived data *)
+    let* ao = Sparse.Ao.open_ao ~mapping_size ~mapping ~data in
+    Sparse.Ao.append_seq_exn ao ~off to_archive;
+    let end_offset = Sparse.Ao.end_off ao in
+    let mapping_end_poff = Sparse.Ao.mapping_size ao in
+    let* () = Sparse.Ao.flush ao in
+    let* () = Sparse.Ao.close ao in
+    (* Prepare new control file *)
+    let start_offset, old_data_end_poff =
+      match t with
+      | Empty _ -> (off, Int63.zero)
+      | Nonempty { control; _ } -> (control.start_offset, control.data_end_poff)
+    in
+    let data_len =
+      Seq.fold_left (fun len str -> len + String.length str) 0 to_archive
+      |> Int63.of_int
+    in
+    let new_control =
+      Control_file.Payload.Volume.V5.
+        {
+          start_offset;
+          end_offset;
+          mapping_end_poff;
+          data_end_poff = Int63.add old_data_end_poff data_len;
+          checksum = Int63.zero;
+        }
+    in
+    (* Write into temporary file on disk *)
+    let tmp_control_path =
+      Irmin_pack.Layout.V5.Volume.control_tmp ~generation ~root
+    in
+    let* c =
+      Control.create_rw ~path:tmp_control_path ~overwrite:true new_control
+    in
+    let* () = Control.fsync c in
```

Comment:
I think it would be fine to remove this. Our usage of `fsync` is a bit haphazard. We do it in places during GC but never (in practice -- there is a flag but it is off for octez) in the main process.